### PR TITLE
Transition to `Integer` trait

### LINKED ISSUE #34: `BoxedUint` support
`crypto-bigint` v0.6.0-pre.0 has a much more featureful `BoxedUint` type than previous releases which hopefully supports all of the prerequisites required for random prime generation: https://docs.rs/crypto-bigint/0.6.0-pre.0/crypto_bigint/struct.BoxedUint.html

We would like to [migrate from `num-bigint-dig` to `crypto-bigint` in the `rsa` crate](https://github.com/RustCrypto/RSA/issues/390), notably because the [current implementation has a timing sidechannel vulnerability](https://github.com/RustCrypto/RSA/issues/19#issuecomment-1822995643). A full migration will require prime generation / primality testing, hence my interest in this issue.

### PR DESCRIPTION
**Pull request by fjarri**
Potentially fixes #34 

In order to add the support for `BoxedUint` we need to make all the methods parametrized by some `UintLike` trait instead of the number of limbs in `Uint`. This PR attempts to see what kind of methods are necessary. See `uint_traits.rs` for what needs to be implemented in `crypto-bigint`. `UintLike` and `UintModLike` can potentially be split into smaller traits. 

Notes:
- There is no `::BITS` anymore, so we need to change the API to get `bit_length: usize` instead of `Option<usize>` everywhere
- Rust does not propagate bounds on `&Self`, which means there are some potentially unnecessary `clone()`s involved, especially in `lucas()` (unless we want to write down scary trait bounds, which we really don't).

### CODE REVIEW

**Comment by tarcieri** [comment]
`UintLike` looks a lot like `crypto_bigint::Integer`, although it will need changes to be able to support dynamically sized types and because of that isn't yet impl'd for `BoxedUint`

**Comment by fjarri** [comment]
Almost, yes, save for `Copy`, `Zero`, associated constants, and a link to the Montgomery representation, which can be a separate trait. I was thinking of finishing up https://github.com/RustCrypto/crypto-bigint/pull/218 while we're at it.

**Comment by tarcieri** [comment]
@fjarri aside from a link to the Montgomery representation, this should take care of the other issues: https://github.com/RustCrypto/crypto-bigint/pull/367

**Comment by fjarri** [comment]
I takes care of some of them, but not all - e.g. vartime ops, bit operations, division, etc are still not in traits. Also it would be nice to have `*Assign` arithmetic operations and arithmetic operations on `&Self`, although it is not critical for the performance.

**Comment by xuganyu96** [comment]
I'd like to help out with the implementation of core functionalities using `BoxedUint`! Is there something that I can get started with?

It seems like there are a lot of functions parameterized by `<L: usize>` that can be transformed into a function that takes an argument `bits_precision: u32`, is that right? I can start implementing them right away.

Thank you!

**Comment by fjarri** [comment]
This PR already includes all the necessary transition to numeric traits instead of `const usize` generic parameters, we just need to move all those traits to `crypto-bigint`. See `uint_traits.rs` for the list of the required functionality.

**Comment by xuganyu96** [comment]
> This PR already includes all the necessary transition to numeric traits instead of `const usize` generic parameters, we just need to move all those traits to `crypto-bigint`. See `uint_traits.rs` for the list of the required functionality.

After pulling this branch it seems that the trait in `uint_traits.rs` is incomplete and this crate will not build. @fjarri @tarcieri if no one else is working on the `UintLike` trait and/or implementing these traits for `Uint/BoxedUint`, I would love to help out!

Thank you!

**Comment by fjarri** [comment]
@tarcieri has been implementing necessary methods for `BoxedUint`, and I have been occupied with other work and hasn't been updating this PR for a while. I was planning to do so after the PRs in https://github.com/RustCrypto/crypto-bigint/issues/268 are merged.

File: Cargo.toml [hunk 1]
```diff
@@ -10,7 +10,8 @@ categories = ["cryptography", "no-std"]
 rust-version = "1.65"
 
 [dependencies]
-crypto-bigint = { version = "0.6.0-pre.0", default-features = false, features = ["rand_core"] }
+#crypto-bigint = { version = "0.6.0-pre.0", default-features = false, features = ["rand_core"] }
+crypto-bigint = { git = "https://github.com/RustCrypto/crypto-bigint.git", default-features = false, features = ["rand_core"], commit = "5ee582b2d1b7077f1e0737210506c53193c24939" }
```
Comment (tarcieri) [inline Cargo.toml:14]: FYI: if you check in Cargo.lock it will pin this for you

**Comment by tarcieri** [comment]
@xuganyu96 the relevant trait in `crypto-bigint` is `Integer`, which is already impl'd for both `Uint` and `BoxedUint` but needs to be expanded with what's missing from `UintLike`: https://docs.rs/crypto-bigint/0.6.0-pre.1/crypto_bigint/trait.Integer.html

**Comment by xuganyu96** [comment]
> @tarcieri has been implementing necessary methods for `BoxedUint`, and I have been occupied with other work and haven't been updating this PR for a while. I was planning to do so after the PRs in [RustCrypto/crypto-bigint#268](https://github.com/RustCrypto/crypto-bigint/issues/268) are merged.

@tarcieri @fjarri  I've taken a try and found the scope of work manageable (see #37 for my progress). If you don't mind I can take over the remainder of the transition (I am a grad student currently on winter break and so I'm looking for work to do anyways). Thank you!

**Comment by fjarri** [comment]
Ok, finally the tests pass. 

Remaining things in `UintLike`:
- `RandomMod` (not included in `Integer`, but perhaps it should be?)
- vartime bit shifts (add new traits in `crypto-bigint`?)
- bit counting methods, vartime and not (add new traits in `crypto-bigint`?)
- some `random_bits(rng, bit_length: u32)` method. Currently there's nothing for that in `Integer`. What should the API be like?
- `TryInto<u32>` or alternatively `TryInto<Limb>`. 

@tarcieri, what's your take on this?

**Comment by tarcieri** [comment]
The reason `Random`/`RandomMod` aren't included in `Integer` is because `rand_core` is currently an optional dependency. We could make it a hard dependency potentially.

Traits for vartime bitshifts sound OK.

There are already various bit counting methods defined on `Integer`, like `bits`, `bits_vartime`, and `trailing_zeros`. Which ones are missing? (e.g. `leading_zeros` I guess?)

> some random_bits(rng, bit_length: u32) method. Currently there's nothing for that in Integer. What should the API be like (e.g. should it return an Option or panic when bit_length is larger than the total bit length in Uint)?

Something like how [`BoxedUint::random`](https://docs.rs/crypto-bigint/0.6.0-pre.5/crypto_bigint/struct.BoxedUint.html#method.random) is defined I guess?

How would `TryInto<Limb>` work?

**Comment by fjarri** [comment]
> There are already various bit counting methods defined on Integer, like bits, bits_vartime, and trailing_zeros. Which ones are missing? (e.g. leading_zeros I guess?)

`trailing_zeros()`, `trailing_ones()`, and `bit_vartime()`

> Something like how [BoxedUint::random](https://docs.rs/crypto-bigint/0.6.0-pre.5/crypto_bigint/struct.BoxedUint.html#method.random) is defined I guess?

Well, that's the question: in case of `Uint` when `bits_precision` is too large, what's the behavior?

**Comment by fjarri** [comment]
On a side note, I think `bits_precision` in `BoxedUint::random()` should be renamed to `bit_length`.

**Comment by fjarri** [comment]
> How would TryInto<Limb> work?

If the number fits in one `Limb`, return that limb, otherwise `None`.

**Comment by tarcieri** [comment]
> Well, that's the question: in case of Uint when bits_precision is too large, what's the behavior?

We'd need a fallible trait/method to cover both cases.

> On a side note, I think bits_precision in BoxedUint::random() should be renamed to bit_length.

It matches `BoxedUint::bits_precision()`. Do you think it should be renamed everywhere?

> If the number fits in one Limb, return that limb, otherwise None.

So, check if the other limbs (if they exist) are zero?

In constant or variable time, and if it's the former, shouldn't it use `CtOption`?

**Comment by fjarri** [comment]
> It matches BoxedUint::bits_precision().

Not really. `BoxedUint::bits_precision()` returns the storage size of `BoxedUint`. the `bits_precision` in `random()` denotes how large the actual number will be, which is different from the storage size.

**Comment by xuganyu96** [comment]
> > There are already various bit counting methods defined on Integer, like bits, bits_vartime, and trailing_zeros. Which ones are missing? (e.g. leading_zeros I guess?)
> 
> `trailing_zeros()`, `trailing_ones()`, and `bit_vartime()`
> 
> > Something like how [BoxedUint::random](https://docs.rs/crypto-bigint/0.6.0-pre.5/crypto_bigint/struct.BoxedUint.html#method.random) is defined I guess?
> 
> Well, that's the question: in case of `Uint` when `bits_precision` is too large, what's the behavior?

`BoxedUint::trailing_zeros`, `BoxedUint::trailing_ones()`, and `BoxedUint::bit_varitme()` have all been implemented and merged in https://github.com/RustCrypto/crypto-bigint/pull/489.




> > It matches BoxedUint::bits_precision().
> 
> Not really. `BoxedUint::bits_precision()` returns the storage size of `BoxedUint`. the `bits_precision` in `random()` denotes how large the actual number will be, which is different from the storage size.
> 
> In fact, `BoxedUint::random()` could have both `bit_length` and (possibly optional) `bits_precision` parameters, to avoid reallocation when widening to the required storage size.

In my attempt [here](https://github.com/entropyxyz/crypto-primes/pull/37), `bit_length` and `bits_precision` indeed need to be separated out into two arguments. This is entirely understandable for `BoxedUint`, but it is a little awkward to use with `Uint` since `bits_precision` is a piece of redundant information that the user will have to fill in. See the code snippet below:

```rust
fn main() {
    // with BoxedUint, separating bit_length and bits_precision is entirely understandable
    // "I want a random number that takes 128 bits to store and 64 bits to represent
    let random_odd = BoxedUint::random(&mut OsRng, 64, 256);

    // with Uint, bits_precision is exactly Uint::BITS, which might be a bit confusing
    // but with clear documentation and some `assert!` I think we can communicate the idea to the users
    let random_odd: U256 = U256::random(&mut OsRng, 64, U256::BITS);
}
```

In practice this means that functions like `generate_prime` will also need to take `bit_length` and `bits_precision` as separate argument regardless of whether the user is using `BoxedUint` or `Uint`:

```rust
let stacked_prime: U256 = generate_prime(&mut OsRng, 64, U256::BITS);
let heaped_prime: BoxedUint = generate_prime(&mut OsRng, 64, 256);
```

File: src/uint_traits.rs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+    fn shr_vartime(&self, shift: u32) -> Self {
+        Self::wrapping_shr_vartime(self, shift)
+    }
+
+    fn shl_vartime(&self, shift: u32) -> Self {
+        Self::wrapping_shl_vartime(self, shift)
+    }
+
+    fn random_bits(rng: &mut impl CryptoRngCore, bit_length: u32) -> Self {
```
Comment (xuganyu96) [inline src/uint_traits.rs:81]: I am sure @fjarri is already aware but this implementation doesn't do what it says to do. In this implementation `random.bits_precision()` returns `bit_length`, so `random >> (bits_precision - bit_length)` will always return `0`.

Comment (fjarri) [reply src/uint_traits.rs:81]: I actually was not aware of it since I only ran tests with `Uint`s. Thanks!

File: src/uint_traits.rs [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        }
+        let random = Self::random(rng);
+        random >> (Self::BITS - bit_length)
+    }
+}
+
+impl UintLike for BoxedUint {
+    fn set_bit_vartime(&mut self, index: u32, value: bool) {
+        if value {
+            *self |= Self::one() << index
```
Comment (xuganyu96) [inline src/uint_traits.rs:75]: `BoxedUint::one` doesn't actually have the same size as `self`. This should probably be replaced with `Self::one_with_precision(self.bits_precision())`

File: src/uint_traits.rs [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    fn bit_vartime(&self, index: u32) -> bool {
+        Self::bit_vartime(self, index)
+    }
+
+    fn sqrt_vartime(&self) -> Self {
+        Self::sqrt_vartime(self)
+    }
+
+    fn overflowing_shl_vartime(&self, shift: u32) -> CtOption<Self> {
+        let (res, is_some) = Self::overflowing_shl(self, shift);
```
Comment (xuganyu96) [inline src/uint_traits.rs:98]: `overflow_shl` returns a truthy `Choice` if `shift` is greater than `bits_precision`. In other words, the choice it returns is true iff the shift overflows. I think you mean this function to return `Some` if the shift does not overflow, so it should probably be like this:

```rust
let (res, overflow) = Self::overflowing_shl(self, shift);
CtOption::new(res, !overflow)
```

Comment (fjarri) [reply src/uint_traits.rs:98]: Yep, you're right, I did that in https://github.com/RustCrypto/crypto-bigint/pull/509

**Comment by fjarri** [comment]
The two uncovered lines in `_is_prime_with_rng()` were uncovered previously. It's hard to find a number that would pass through the first MR test and the Lucas test, but not the second MR test. And the branch that returns `Prime` from the Lucas test is inactive because all such numbers were already sieved out at the `Sieve` stage.

**Pull request merged**